pkg/fleetallocation/find.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fleetallocation
+
+import (
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
+)
+
+// nodeCount is just a convenience data structure for
+// keeping relevant GameServer counts about Nodes
+type nodeCount struct {
+	ready     int64
+	allocated int64
+}
+
+// findReadyGameServerForAllocation is a O(n) implementation to find a GameServer with priority
+// defined in the comparator function.
+func findReadyGameServerForAllocation(gsList []*v1alpha1.GameServer, comparator func(bestCount, currentCount *nodeCount) bool) *v1alpha1.GameServer {
+	counts := map[string]*nodeCount{}
+	// track potential gameservers, one for each node
+	allocatableGameServers := map[string]*v1alpha1.GameServer{}
+
+	// count up the number of allocated and ready game servers that exist
+	// also, since we're already looping through, track one Ready GameServer
+	// per node, so we can use that as a short list to allocate from
+	for _, gs := range gsList {
+		if gs.DeletionTimestamp.IsZero() &&
+			(gs.Status.State == v1alpha1.Allocated || gs.Status.State == v1alpha1.Ready) {
```

Comment:
I'm expecting a lot of allocations so we might want to find a way latter to cache that informations or have a single source of truth instead of traversing the list. Food for thoughts.

Reply:
Yeah, I was debating that too - maybe we keep a running count for each node at all times - but I was concerned that would add complexity. But definitely food for thought, and potentially an optimisation as we build out.